### PR TITLE
Add flatten op interface and enhance APIs about detection to support variable-length image.

### LINKED ISSUE #12304: 使用Fluid对图像变长数据多次训练/预测存在的问题
Fluid构建一个网络分为两个过程：`compile time`和`run time`
1. 对于变长数据的预测过程可以分为以下两步：
`Compile time`过程Fluid根据Python输入data_const的shape计算得到`ProgramDesc`,根据输入data的shape计算得到各个Var的shape
`Run time`根据实际feed的data_real进行计算，如果输入是变长数据，data_real与data_const的shape不同，可能会引起一些op的错误
2. 已经发现的问题
目标检测相关的一些Reshape API使用的是输入`Compile time`的shape，所以在输入变长数据时会出错
ReshapeOp相关的一些：
https://github.com/PaddlePaddle/Paddle/blob/9e0a94f069329b430d0843327e474712e079a660/python/paddle/fluid/layers/detection.py#L266-L268
https://github.com/PaddlePaddle/Paddle/blob/9e0a94f069329b430d0843327e474712e079a660/python/paddle/fluid/layers/detection.py#L679-L680
https://github.com/PaddlePaddle/Paddle/blob/edc32c089605944eda860e3404f8aff152a4ff38/python/paddle/fluid/layers/detection.py#L691
https://github.com/PaddlePaddle/Paddle/blob/9e0a94f069329b430d0843327e474712e079a660/python/paddle/fluid/layers/detection.py#L1009-L1012
https://github.com/PaddlePaddle/Paddle/blob/edc32c089605944eda860e3404f8aff152a4ff38/python/paddle/fluid/layers/detection.py#L1102-L1106
https://github.com/PaddlePaddle/Paddle/blob/9e0a94f069329b430d0843327e474712e079a660/python/paddle/fluid/layers/detection.py#L1118-L1122

### PR DESCRIPTION
**Pull request by baiyfbupt**
fix #12304

### CODE REVIEW

File: doc/fluid/api/layers.rst [hunk 1]
```diff
@@ -1776,3 +1776,11 @@ rank_loss
...
     :noindex:
 
+.. _api_fluid_layers_flatten:
+
+flatten
+-------
+
+..  autofunction:: paddle.fluid.layers.flatten
+    :noindex:
+
```
Comment (qingqing01) [inline doc/fluid/api/layers.rst]: It seems,  the new doc doest not depend on `doc/fluid/api/layers.rst`. So remove this.

File: python/paddle/fluid/layers/detection.py [hunk 2]
```diff
@@ -262,10 +263,11 @@ class number, M is number of bounding boxes. For each category
...
         target_box=loc,
         code_type='decode_center_size')
-    old_shape = scores.shape
-    scores = nn.reshape(x=scores, shape=(-1, old_shape[-1]))
+    compile_shape = scores.shape
+    run_shape = ops.shape(scores)
+    scores = nn.flatten(x=scores, axis=2)
     scores = nn.softmax(input=scores)
-    scores = nn.reshape(x=scores, shape=old_shape)
+    scores = nn.reshape(x=scores, shape=compile_shape, actual_shape=run_shape)
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py:273]: shape 和 actual_shape不是只用其中一个就可以了吗？这里不需要 `compile_shape`吧？

File: python/paddle/fluid/layers/detection.py [hunk 3]
```diff
@@ -688,7 +690,10 @@ def __reshape_to_2d(var):
 
     # 2. Compute confidence for mining hard examples
     # 2.1. Get the target label based on matched indices
-    gt_label = nn.reshape(x=gt_label, shape=gt_label.shape + (1, ))
+    gt_label = nn.reshape(
+        x=gt_label, shape=(len(gt_label.shape) - 1) * (0, ) + (
+            -1,
+            1, ))
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py]: 这里为啥这样改？

File: python/paddle/fluid/layers/detection.py (hunk 2 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/detection.py:273]: 需要shape attr在compile time构建图
https://github.com/PaddlePaddle/Paddle/blob/02c31458bbb84155ba46d362c9c61af4b36902b4/paddle/fluid/operators/reshape_op.cc#L36-L37

Comment (qingqing01) [reply python/paddle/fluid/layers/detection.py:273]: 这里之前就讨论过，是`reshape_op`内部实现的问题，需要增强reshape_op内部实现。

否则，按照现在这样，用户使用Python接口，很疑惑为啥需要指定2个，非常奇怪的形式。

Comment (baiyfbupt) [reply python/paddle/fluid/layers/detection.py:273]: 现在reshape_op是这样设计的，确实有点麻烦，需要增强一下
https://github.com/PaddlePaddle/Paddle/blob/02c31458bbb84155ba46d362c9c61af4b36902b4/paddle/fluid/operators/reshape_op.cc#L172-L174

**Review by qingqing01 — COMMENTED** [review]
LGTM. But need to update API.spec

File: python/paddle/fluid/layers/detection.py (hunk 3 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/detection.py]: 不需要这么复杂，已改正

**Review by qingqing01 — DISMISSED** [review]
LGTM.

**Comment by qingqing01** [comment]
@panyx0718  @typhoonzero  This PR add a Python API interface, please help to review. Thanks!

**Review by typhoonzero — DISMISSED** [review]
LGTM for API change

File: python/paddle/fluid/layers/nn.py [hunk 4]
```diff
@@ -5354,3 +5355,70 @@ def rank_loss(label, left, right, name=None):
...
+    Case 2:
+      Given
+        X.shape = (3, 100, 100, 4)
+      and
+        axis = 0
+      We get:
+        Out.shape = (1, 3 * 100 * 100 * 4)
+
+    Args:
+        x (Variable): A tensor of rank >= axis.
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:5390]: is axis zero-based?

**Comment by baiyfbupt** [comment]
@panyx0718 @typhoonzero @qingqing01 Please help to review again. Many thanks! This PR add a Python API interface.

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -5354,3 +5355,70 @@ def rank_loss(label, left, right, name=None):
...
+        axis (int): Indicate up to which input dimensions (exclusive) should 
+                    be flattened to the outer dimension of the output. 
+                    The value for axis must be in the range [0, R], where R
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:5393]: should be [0, R)?

File: python/paddle/fluid/layers/nn.py [hunk 6]
```diff
@@ -5354,3 +5355,70 @@ def rank_loss(label, left, right, name=None):
...
+    Flattens the input tensor into a 2D matrix.
+
+    Examples:
+    Case 1:
+      Given
+        X.shape = (3, 100, 100, 4)
+      and
+        axis = 2
+      We get:
+        Out.shape = (3 * 100, 4 * 100)
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:5379]: why is this not (3, 100, 4 * 100)?

File: python/paddle/fluid/layers/nn.py [hunk 7]
```diff
@@ -5354,3 +5355,70 @@ def rank_loss(label, left, right, name=None):
                 "Right": right},
         outputs={'Out': out})
     return out
+
+
+def flatten(x, axis=1, name=None):
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:5367]: what is this api design based on? Seems not the same as numpy.
https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.ndarray.flatten.html

File: python/paddle/fluid/layers/nn.py (hunk 4 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/nn.py:5390]: Yes

File: python/paddle/fluid/layers/nn.py (hunk 5 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/nn.py:5393]: Same as onnx https://github.com/onnx/onnx/blob/master/docs/Operators.md#flatten : When axis=R, output_tensor is `((d_0 X d_1 ... d_n), 1)`

File: python/paddle/fluid/layers/nn.py (hunk 6 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/nn.py:5379]: output is a 2D tensor,`(d_0 X d_1 ... d_(axis-1), d_axis X d_(axis+1) ... X dn)`

File: python/paddle/fluid/layers/nn.py (hunk 7 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/nn.py:5367]: based on onnx operator Flatten

**Comment by panyx0718** [comment]
there is a swp file

**Comment by qingqing01** [comment]
Please remove `paddle/fluid/operators/.flatten_op.cc.swp`

**Comment by baiyfbupt** [comment]
@panyx0718 @qingqing01 `.flatten_op.cc.swp` was added in previous PR by mistake, this PR is trying to remove it.

**Review by qingqing01 — APPROVED** [review]

**Review by panyx0718 — APPROVED** [review]
Approve API

**Pull request merged**